### ISSUE TITLE
Supporting iceman fork

### ISSUE THREAD

**Issue by micolous** [AndProx/AndProx#12]
Lots of people are using iceman's fork.  This has a protocol that is incompatible with the upstream client, and has extra functionality that is not in the upstream client.

I need to build another APK which is based on Iceman's client, and come up with some way so that it is easier to pull in other forks in the future.

**Note:** This issue is currently blocked on getting my serial communication patches into upstream PM3 (and Iceman): Proxmark/proxmark3#463

This is progressing well as of April-May 2018.

**Comment by micolous** [comment]
In order to get this working on Iceman's fork, this requires the following PRs to be applied:

- https://github.com/Proxmark/proxmark3/pull/506 (Externalise PrintAndLog)
- https://github.com/Proxmark/proxmark3/pull/587 (Comms refactor 1)
- https://github.com/Proxmark/proxmark3/pull/595 (Comms refactor 2)
- https://github.com/Proxmark/proxmark3/pull/596 (Comms refactor 3)
- https://github.com/Proxmark/proxmark3/pull/607 (Comms refactor 4)
- https://github.com/Proxmark/proxmark3/pull/625 (iceman fork doesn't unlink, but if refactor picks up that patch chain...)

Nice to have:

- Externalise PrintAndLogEx (because having structured data about the log level would be nice)

**Comment by pwpiwi** [comment]
> (iceman fork doesn't unlink, but if refactor picks up that patch chain...)

Oh yes, it does in flasher.c:
```
void CloseProxmark() {
	// Clean up the port
	uart_close(sp);
	// Fix for linux, it seems that it is extremely slow to release the serial port file descriptor /dev/*
	unlink(serial_port_name);
}
```

**Comment by micolous** [comment]
Good catch! But I'm  not too worried about the flasher, as I'd just use the
mainline flasher.

**Comment by ceres-c** [comment]
Hi, do you have any update on Iceman fork support? Are you still working on it?
I'd be very much interested...

**Comment by micolous** [comment]
@ceres-c This issue is in the "backlog".  I'll be working on getting **a** release with mainline PM3 working before I work on any other forks.

All of the patches from https://github.com/AndProx/AndProx/issues/12#issuecomment-408039310 will need to be in iceman (or any other fork) in order to make it supportable. If you want to help, getting those patches / refactorings into iceman's fork would be helpful.

For a benchmark, it took 8 months to get all of my patches into mainline PM3.  Please expect this to take at least that long if I'm doing this solo.

**Comment by pwpiwi** [comment]
Patches usually make it much faster into the Iceman fork.
